---
title: "class_anal"
output: pdf_document
date: "2024-04-04"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
require(tidyverse)
require(knitr)
require(kableExtra)
require(cowplot)
require(ggpubr)
```

```{r cars}
# locations
out_dir <- "~/Desktop/Current_Classes/Applied_Bioinformatics/proj_stats_full/"
stats <- "~/Desktop/Current_Classes/Applied_Bioinformatics/proj_stats_full/"

# reading in the data
df <- read.csv2(paste0(stats, "out.csv"), sep=",")
df <- df |> 
    mutate(X.rname = as.factor(X.rname),
           meandepth = as.double(meandepth))

# adding in read bp stats
read_bp <- read.csv2(paste0(out_dir, "Mmelodia_file_numbered_reads"), sep=" ", header = FALSE)
read_bp$V2 <- substr(read_bp$V2, 1, nchar(read_bp$V2) - 1)
read_bp <- read_bp |> 
    select(V2, V3)
names(read_bp) <- c("Filename", "read_bp")

# adding read bp to data + removing shitty raw read data
df <- right_join(df, read_bp, by = "Filename") |> 
    filter(Filename != "L1_S29_2471-03658_Mmelodia_PonquogueBeach-NY_2022") |> 
    filter(Filename != "L1_S25_2831-61178_Mmelodia_GreatBayBlvdWMA-NJ_2021")

# make the coverage targets in order
level_order = c("cov_1","cov_3","cov_5","cov_7","cov_10","cov_15","cov_20","full")
df$coverage_target <- factor(df$coverage_target, levels = level_order)
```

```{r}
cov_tab <- df |>     
    select(Filename, read_Count, scaffold_type, read_bp) |> 
    filter(scaffold_type == "NC") |> 
    group_by(Filename) |> 
    summarise(cov = floor((mean(read_Count) / 2) / ifelse(mean(read_bp) == 150, 2545000, 1592000)),
              mean_rc = mean(read_Count) / 2,
              num_used = ifelse(mean(read_bp) == 150, 2545000, 1592000)) 


new_lines <- data.frame(Filename = c(),
                        Reads = c(),
                        Coverage = c())

# Iterate over each row in the dataframe
for (i in 1:nrow(cov_tab)) {
  # Extract values from the current row
  name <- cov_tab$Filename[i]
  cov_to_seq <- cov_tab$cov[i]
  num_read <- cov_tab$num_used[i]
  
  sequence <- list(1, 3, 5, 7, 10, 15, 20)
  sequence  <- sequence[sequence <= cov_to_seq]
  # Create the sequence
  
  for (j in sequence) {
      new_row <- data.frame(Filename = c(name),
                            Reads = c(j * num_read),
                            Coverage = c(j))
      new_lines <- rbind(new_lines, new_row)
  }
}

#new_lines$Filename <- paste0(new_lines$Filename, "_")

#write.table(new_lines, file = "Mmelodia_slurm_input_with_reads", sep = " ", col.names = FALSE, row.names = TRUE, quote = FALSE)


kable(cov_tab) |> 
 kable_styling(bootstrap_options = "bordered") |>
 as_image(file = paste0(out_dir, "cov.png"))
```


```{r, echo=FALSE}
by_file <- df |> 
    ggplot(aes(x=scaffold_type, y=meandepth)) +
        geom_boxplot() +
        facet_wrap(vars(Filename)) +
        ylim(0, 150)
by_file

by_scaffold_type <- df |> 
  filter(coverage_target == "full") |> 
  #filter(scaffold_type == "NC") |> 
    ggplot(aes(x=X.rname, y=meandepth, fill=scaffold_type)) +
        geom_boxplot(outlier.size=0.1, outlier.stroke = 0.5, linewidth=0.1) +
        ylim(0, 150) +
        labs(x="Scaffolds", 
             y="Mean Depth",
             fill="Scaffold Type") +
        #theme(axis.text.x = element_blank()#element_text(angle = 90, size = 2),
        #     ) +
        theme(axis.text.x = element_blank()) +#element_text(angle = 90)) +
        scale_fill_manual(values = c("forestgreen", "pink"))
by_scaffold_type
ggsave(filename=paste0(out_dir, "scaffolds_full.png"), width=10, height=3, units="in", dpi="retina")
#ggsave(filename=paste0(out_dir, "scaffolds_full_NC.png"), width=10, height=3, units="in", dpi="retina")
```

```{r}
depth_table_by_scaffold <- df |> 
    select(X.rname, meandepth) |> 
    group_by(X.rname) |> 
    summarise(mean_depth = mean(meandepth))

kable(depth_table_by_scaffold)

depth_table_total_by_scaffold_type <- df |> 
    select(scaffold_type, meandepth) |> 
    group_by(scaffold_type) |> 
    summarise(mean_depth = mean(meandepth))

kable(depth_table_total_by_scaffold_type)

depth_table_by_file_scaffold_type <- df |> 
    select(Filename, scaffold_type, meandepth) |> 
    group_by(Filename, scaffold_type) |> 
    summarise(mean_depth = mean(meandepth))

out <- "mean_depth_by_file.png"

#kable(depth_table_by_file_scaffold_type) |> 
#    kable_styling(bootstrap_options = "bordered") |>
#   as_image(file = out)
#mean(df$"meandepth")
```

```{r}
one_X_coverage <- df |> 
    select(Filename, meandepth, read_Count, scaffold_type, read_bp, coverage_target) |>
    group_by(Filename, scaffold_type, read_bp, coverage_target) |> 
    summarise(one_X_cov_med = median(read_Count) / median(meandepth),
              one_X_cov_mean = mean(read_Count) / mean(meandepth),
              read_Count_median = median(read_Count),
              meandepth_median = median(meandepth),
              read_Count_mean = mean(read_Count),
              meandepth_mean = mean(meandepth))
              

one_x_tab <- one_X_coverage |> 
    group_by(scaffold_type, read_bp) |> 
    summarise(one_x_median_per_fastq = median(one_X_cov_med) / 2,
              one_x_mean_per_fastq = mean(one_X_cov_mean) / 2) 

one_x_tab |> 
  kable() |> 
  kable_styling(bootstrap_options = "bordered") |> 
  as_image(file = paste0(out_dir, "one_x_read_count_table.png"))

```

```{r}
df |> 
    select(coverage, meandepth, scaffold_type, coverage_target) |> 
    group_by(scaffold_type, coverage_target) |> 
    summarise(mean_depth = mean(meandepth)) |> 
    kable() |> 
    kable_styling(bootstrap_options = "bordered") |>
    as_image(file = paste0(out_dir, "actual_coverage_calcs_table.png"))
  
```


```{r}
one_X_coverage$coverage_target_real <- NA
for (i in 1:nrow(one_X_coverage)) {
  if (one_X_coverage$coverage_target[i] != "full") {
  one_X_coverage$coverage_target_real[i] <- as.integer(sub("^cov_(.*)$", "\\1", one_X_coverage$coverage_target[i])) / 2
  } else {
     one_X_coverage$coverage_target_real[i] <- "full"
  }
}
level_order = c("0.5","1","1.5","2.5","3.5","5","7.5","10","full")
one_X_coverage$coverage_target_real <- factor(one_X_coverage$coverage_target_real, levels = level_order)

l <- labeller(coverage_target_real=c("0.5"="0.5x Coverage",
                                     "1"="1x Coverage",
                                     "1.5"="1.5x Coverage",
                                     "2.5"="2.5x Coverage",
                                     "3.5"="3.5x Coverage",
                                     "5"="5x Coverage",
                                     "7.5"="7.5x Coverage",
                                     "10"="10x Coverage",
                                     "full"))
    
one_X_coverage |> 
    filter(coverage_target != "full") |> 
    select(Filename, one_X_cov_mean, scaffold_type, coverage_target_real) |> 
    ggplot(aes(x=Filename, y=one_X_cov_mean/1e7, color=scaffold_type)) +
        geom_point() +
        theme_minimal() +
        labs(x="File", y="Count of Reads in 1X coverage (10 Millions)", color="Scaffold Type") +
        theme(#axis.text.x = element_text(angle=90, vjust=0.5, hjust=1, size=1),
              axis.text.x = element_blank(),
              #axis.title.y = element_text(size=8),
              #axis.title.x = element_text(size=8)
              ) +
        scale_color_manual(values = c("forestgreen", "pink")) +
        facet_wrap(~ coverage_target_real, ncol = 4, labeller = l)
#ggsave(filename=paste0(out_dir, "one_X_coverage_full.png"), width=5, height=3, units="in", dpi="retina")
ggsave(paste0(out_dir, "one_X_coverage_all.png"), height = 8, width = 17)

#kable(one_X_coverage) |> 
#    kable_styling(bootstrap_options = "bordered") |>
#    as_image(file = "one_X_coverage_by_file.png")
```


```{r}
df |>
    #filter(coverage_target == "full") |> 
    filter(coverage_target != "full") |>
    select(meandepth, read_Count, Filename, scaffold_type, coverage_target) |> 
    group_by(Filename, scaffold_type) |> 
    summarise(read_count = median(read_Count) / 1e8,
              median_depth = median(meandepth)) |> 
    ggplot(aes(x=read_count, y=median_depth, color=scaffold_type)) +
        geom_point() +
        labs(x="Median Read Count (100 millions)", y="Median Scaffold Depth", color="Scaffold Type") +
        geom_line(aes(group=Filename), color = "black", alpha = 0.5) +
        geom_smooth(method = "lm", se=FALSE, size=0.5) +
        stat_cor(aes(label = ..rr.label..), label.x = 0.1) +
        #stat_regline_equation() +
        theme_minimal() +
        scale_color_manual(values = c("forestgreen", "pink"))
ggsave(paste0(out_dir, "depth_by_read_count_full.png"), height = 5, width = 7)
```

```{r}
unfiltered <- "~/Desktop/Current_Classes/Applied_Bioinformatics/sparrow_study/data/snps_unfiltered.csv"
filtered <- "~/Desktop/Current_Classes/Applied_Bioinformatics/sparrow_study/data/snps.csv"
snps_filtered <- read_delim(filtered, delim=" ", col_names=FALSE)
snps_filtered$status <- "filtered"
names(snps_filtered) <- c("file", "snps", "status")
snps_unfiltered <- read_delim(unfiltered, delim=" ", col_names=FALSE)
snps_unfiltered$status <- "unfiltered"
names(snps_unfiltered) <- c("file", "snps", "status")
snps <- rbind(snps_filtered, snps_unfiltered)
rm(snps_filtered, snps_unfiltered)
#cov_10/bcf_out/2551-18714_Mmelodia_ScarboroughMarshME_20160731_complete.bcf.gz
snps$coverage <- sub("^(.*)/bcf_out.*$", "\\1", snps$file)
snps <- snps |> 
  filter(coverage != "full")

levs <- c("cov_1", "cov_3", "cov_5", "cov_7", "cov_10", "cov_15", "cov_20")
coverage_list <- c("cov_1"="0.5x", "cov_3"="1.5x", "cov_5"="2.5x", "cov_7"="3.5x", "cov_10"="5x", "cov_15"="7.5x", "cov_20"="10x")
snps$coverage <- factor(snps$coverage, levels=levs, labels=coverage_list)

snps$file <- sub("^.*/bcf_out/(.*)_complete.bcf.gz$", "\\1", snps$file)
```

```{r}

snps_anim <- snps |> 
  filter(coverage != "full") |> 
  ggplot(aes(y=snps/1000000, x=file, color=status)) +
    geom_point() +
    geom_line(aes(group=file), color = "black", alpha = 0.5) +
    labs(x="file",
         y="SNPs (millions)",
         color="Filtered Status") +
    #facet_wrap(~coverage) +
    theme_minimal_grid() +
    scale_color_manual(values=c("forestgreen", "pink")) +
    theme(axis.text.x = element_blank()) +
    ggtitle('Coverage: {closest_state}') +
    transition_states(states = coverage, transition_length = 3, state_length = 5) +
    exit_recolor(color = "black") +
    exit_fade(alpha=0.5)

animate(snps_anim, fps=20, duration=10, res=150, height=3, width=5, units="in")
anim_save("~/Desktop/Current_Classes/Applied_Bioinformatics/sparrow_study/data/SNPS_anim.gif")
```



```{r, fig.height=30, fig.width=10}
depth_table_by_file_scaffold_type |> 
    ggplot(aes(x=scaffold_type, y=mean_depth, fill=scaffold_type)) +
        geom_col() +
        coord_flip() +
    
        facet_wrap(vars(Filename), ncol=1, strip.position = "left") +
        theme_minimal() +
        theme(strip.text.y.left = element_text(angle = 0),
              strip.background.y = element_blank(),
              axis.text.y = element_blank(),
              axis.ticks.y = element_blank(),
              plot.background = element_rect(fill = "white")) +
        scale_fill_manual(values = c("forestgreen", "pink"))
ggsave("depth_by_file.png", height = 30, width = 10)
```

```{r, fig.height=30, fig.width=5}
df |> 
    select(X.rname, meandepth, scaffold_type) |> 
    ggplot(aes(x=X.rname, y=meandepth, fill=as.factor(scaffold_type))) +
        geom_boxplot() +
        coord_flip() +
        #ylim(0, 150) +
        theme_minimal() +
        scale_fill_manual(values = c("forestgreen", "pink"))   
ggsave("depth_by_scaffold.png", height = 30, width = 10)
```

```{r}
depth_table_by_file_scaffold_type |> 
    ggplot(aes(x=mean_depth, fill=scaffold_type)) +
        geom_histogram(alpha=0.750, position = "identity") +
        theme_minimal() +
        scale_fill_manual(values = c("forestgreen", "pink"))
ggsave("hist_depths.png", height = 5, width = 7)
```


